crates/biome_js_type_info/src/resolver.rs

```diff
@@ -17,81 +62,360 @@ const MAX_RESOLUTION_DEPTH: usize = 16;
 /// trait.
 pub trait TypeResolver {
-    /// Resolves a type by its `qualifier`.
-    fn resolve_qualifier(&self, qualifier: &TypeReferenceQualifier) -> Option<Type>;
+    /// Returns the resolution level handled by this resolver.
+    fn level(&self) -> TypeResolverLevel;
 
-    /// Resolves the type of a value by its `identifier`.
-    fn resolve_type_of(&self, identifier: &Text) -> Option<Type>;
-}
+    /// Attempts to find the given type data if it already exists within the
+    /// resolver.
+    fn find_type(&self, type_data: &TypeData) -> Option<TypeId>;
 
-impl Type {
-    /// Attempts to resolve all unresolved types using the given resolver.
-    pub fn resolve(&mut self, resolver: &dyn TypeResolver) {
-        let stack = &[&**self];
-        *self = if self.needs_resolving(resolver, stack) {
-            self.resolved(resolver, stack)
-        } else {
-            self.flattened(resolver, stack)
-        };
+    /// Returns type data by its ID.
+    ///
+    /// **Warning:** The given [`TypeId`] must be a type ID that was created by
+    ///              this resolver.
+    fn get_by_id(&self, id: TypeId) -> &TypeData;
+
+    /// Returns type data by its resolved ID, if possible.
+    fn get_by_resolved_id(&self, id: ResolvedTypeId) -> Option<&TypeData>;
+
+    /// Returns the [`TypeReference`] to refer to a [`TypeId`] belonging to this
+    /// resolver.
+    fn reference_to_id(&self, id: TypeId) -> TypeReference {
+        TypeReference::Resolved(ResolvedTypeId(self.level(), id))
     }
-}
 
-/// Trait to be implemented by `Type` and its subtypes to aid the resolver.
-pub(crate) trait Resolvable {
-    /// Returns whether (part of) the type can and should be resolved.
-    fn needs_resolving(&self, resolver: &dyn TypeResolver, stack: &[&TypeInner]) -> bool;
+    /// Returns a reference to the given type data, if possible.
+    fn reference_to_data(&self, type_data: &TypeData) -> Option<TypeReference> {
+        match type_data {
+            TypeData::Reference(reference) => Some(reference.as_ref().clone()),
+            other => self.find_type(other).map(|id| self.reference_to_id(id)),
+        }
+    }
 
-    /// Returns the resolved version of this type.
+    /// Returns a reference to the given type data, registering the data if
+    /// necessary.
+    fn reference_to_registered_data(&mut self, type_data: TypeData) -> TypeReference {
+        match self.reference_to_data(&type_data) {
+            Some(type_data) => type_data,
+            None => {
+                let id = self.register_type(type_data);
+                self.reference_to_id(id)
+            }
+        }
+    }
+
+    /// Registers a type within the level handled by this resolver.
     ///
-    /// You should first call `Self::needs_resolving()` to avoid making
-    /// needless clones.
-    fn resolved(&self, resolver: &dyn TypeResolver, stack: &[&TypeInner]) -> Self;
-}
+    /// Should perform automatic deduplication if a type with the given
```

Comment:
What does "deduplication" mean in this context?

Reply:
Reworded it a bit. It basically meant if you tried to register a type that was already there, it would return an existing `TypeId`.